### PR TITLE
sha256 check sums to verify original and converted model data

### PR DESCRIPTION
**Pull request by gjmulder**
Not a developer, so my git-fu is a bit rusty. Hopefully this pull request covers everything?!?

* Add shadow `./model.sha256` dir containing a dir for each model and a corresponding `checklist.sha256` containing sha256 sums of the *.pth *bin* and *json files

* Add script `./model.sha256/chk_sha256sums.sh` to walk user-supplied `./models` subdir and run sha256sum against above files to diff checklist.sha256 for each model

* Update` README.md` with corresponding instructions

### CODE REVIEW

**Comment by prusnak** [comment]
Good idea, but I have one suggestions.

How about we put all hashes into a single file `models/SHA256SUMS`?

That's more standard way how to do it and you can then perform a single check using:

```
sha256sum -c models/SHA256SUMS
```

on Linux, or on macOS:

```
shasum -a 256 -c models/SHA256SUMS
```

That way you can drop the shellscript completely.

**Comment by gjmulder** [comment]
@prusnak Thx! RTFM on `sha256sum` would have saved me an hour of unnecessary scripting. :facepalm:

I'm not sure if there's a directory naming convention for the alpaca model subdirs, so I am proposing one.

**Comment by prusnak** [comment]
> I put the `SHA256SUMS` in the root dir as `models` is user-provided

Great!

Can you please sort the file using `sort SHA256SUMS > foo ; mv foo SHA256SUMS` and update the PR?

File: README.md [hunk 1]
```diff
@@ -224,6 +224,14 @@ cadaver, cauliflower, cabbage (vegetable), catalpa (tree) and Cailleach.
 > 
 ```
 
+### Obtaining and verifying the Facebook LLaMA original model and Stanford Alpaca model data
+
+* The LLaMA models are officially distributed by Facebook and will never be provided through this repository. See this [pull request in Facebook's LLaMA repository](https://github.com/facebookresearch/llama/pull/73/files) if you need to obtain access to the model data.
+* Please verify the sha256 checksums of all of your `consolidated*.pth` and corresponding converted `ggml-model-*.bin` model files to confirm that you have the correct model data files before logging an issue relating to your model files.
+
+The following command will verify if you have all possible latest files in your self-installed `./models` subdirectory:
+`llama.cpp$ sha256sum -c SHA256SUMS`
```
Comment (prusnak) [inline README.md:233]:
```suggestion
`sha256sum -c SHA256SUMS` on Linux
or
`shasum -a 256 -c SHA256SUMS` on macOS
```

**Comment by mqy** [comment]
Good job, thanks a lot :)

I personally recommend renaming `SHA256SUMS` as `models.sha256`, with the following reasons:

* Prefix with `models`, so it's clear that this file belongs to models.
* Just like `.sha1`, it's clear that the `.sha256` file contains sh256 checksum(s).
* Prefer lower case if possible, because upper case is a bit hard to read.

**Comment by prusnak** [comment]
The filename `SHA256SUMS` is pretty standard and people are used to it. Examples:
- https://releases.ubuntu.com/22.10/
- https://bitcoincore.org/bin/bitcoin-core-24.0.1

Also later we can add more hashes to the file, not strictly related to the models, so naming it `models.sha256` seems wrong.

**Comment by sw** [comment]
Can you add `--ignore-missing` to the `sha256sum` command line in the readme? Many people won't have all files, the output is less noisy that way. (This is for Linux, don't know about MacOS).

I had a check failure with models/7B/ggml-model-q4_0.bin, I think a recent commit may have lead to some floating point rounding differences?

**Comment by gjmulder** [comment]
> Can you add `--ignore-missing` to the `sha256sum` command line in the readme? Many people won't have all files, the output is less noisy that way. (This is for Linux, don't know about MacOS).
> 
> I had a check failure with models/7B/ggml-model-q4_0.bin, I think a recent commit may have lead to some floating point rounding differences? I have b85058443e89dabdf674d5018d979f0d682977f8413f05b5fd235d36d7a8ff82 for that file.

@sw I just regenerated everything from the *.pth files and now our checksums agree:

```
llama.cpp$ git log | head -1
commit 353ec251a42491f5192c48561da4b444ef67f23c
llama.cpp$ grep b85058443e89dabdf674d5018d979f0d682977f8413f05b5fd235d36d7a8ff82 SHA256SUMS 
b85058443e89dabdf674d5018d979f0d682977f8413f05b5fd235d36d7a8ff82  models/7B/ggml-model-q4_0.bin
```

**Comment by prusnak** [comment]
> (This is for Linux, don't know about MacOS).

Yes, it works for macOS too. I updated my suggestion here to contain the option: https://github.com/ggerganov/llama.cpp/pull/338#discussion_r1143489500

**Comment by prusnak** [comment]
I went ahead and implemented the suggestions from above and rebased/squashed on top of the current master.

**Review by prusnak — APPROVED** [review]

**Comment by prusnak** [comment]
Thanks @gjmulder for computing the hashes. Merged!

**Comment by anzz1** [comment]
Not all of these checksums seem to be correct. Are they calculated with the "v2" new model format after the tokenizer change? PR: https://github.com/ggerganov/llama.cpp/pull/252 Issue: https://github.com/ggerganov/llama.cpp/issues/324 

For example, "models/alpaca-7B/ggml-model-q4_0.bin"

v1: 1f582babc2bd56bb63b33141898748657d369fd110c4358b2bc280907882bf13
v2: 8d5562ec1d8a7cfdcf8985a9ddf353339d942c7cf52855a92c9ff59f03b541bc 

The SHA256SUMS file has the old v1 hash.

**Comment by mqy** [comment]
I can confirm `ggml-model-q4*` in 7B and 13B  mismatch that in SHA256SUMS:

- b85058443e89dabdf674d5018d979f0d682977f8413f05b5fd235d36d7a8ff82  models/7B/ggml-model-q4_0.bin
- d3ab1548a2d19d989c1e7ea1130ab8c6300c75941a434a2d333ef564f989131b  models/13B/ggml-model-q4_0.bin
- 38b705ce6c5baba4bb6056f11189a4ad21b70591258422450e30495b6ccd8521  models/13B/ggml-model-q4_0.bin.1

Other files are correct. I regenerated mismatched q4 files with latest program (make clean; make).

**Comment by prusnak** [comment]
Please open new pull requests if something is wrong.

**Pull request merged**